webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,219 @@
+  // syncStarted updates when recieving msg of {sync_id, complated: false}
+  #syncStarted: string | undefined = undefined;
+  // syncStarted updates when recieving msg of {sync_id, complated: true}
+  #syncComplete: string | undefined = undefined;
+  // List of messages recieved from server
+  #pendingMsg: Array<Record<string, StreamContent>> = [];
+
+  //callbacks
+  #onUpsert: (m: Record<string, StreamContent>) => void;
+  #onDelete: (s: Streamable, a: Array<number>) => void;
+  #isLoading: ((b: boolean) => void) | undefined;
+
+  constructor(
+    wsUrl: string,
+    onUpsert: (m: Record<string, StreamContent>) => void,
+    onDelete: (s: Streamable, a: Array<number>) => void,
+    isLoading?: (b: boolean) => void,
+  ) {
+    this.#wsUrl = wsUrl;
+    this.#onUpsert = onUpsert;
+    this.#onDelete = onDelete;
+    this.#isLoading = isLoading;
+    this.#advance();
+  }
+
+  #connect() {
+    this.#ws = new WebSocket(this.#wsUrl);
+    this.#ws.onopen = () => {
+      logger.info('Streaming websocket opened!');
+      this.#retries = 0;
+      this.#advance();
+    };
+    this.#ws.onerror = (err) => {
+      // No need to do anything else becauses onerror will trigger onclose
+      logger.error('Streaming websocket errored: ', err);
+    };
+
+    this.#ws.onclose = () => {
+      this.#syncSent = undefined;
+      const backoff = backoffs[this.#retries];
+      if (backoff === undefined) {
+        throw new Error('Websocket cannot reconnect!');
+      }
+      this.#timeout = dayjs().add(backoff, 'second');
+      this.#retries += 1;
+      logger.info(`#${this.#retries} of retries: in ${backoff}s`);
+      setTimeout(() => this.#advance(), backoff * 1000);
+    };
+
+    this.#ws.onmessage = (event) => {
+      const msg = JSON.parse(event.data) as Record<string, StreamContent>;
+      this.#pendingMsg.push(msg);
+      this.#advance();
+    };
+  }
+
+  #shouldSkip(newSub: Record<Streamable, Subscription>): boolean {
+    if (!this.#curSub) return false;
+    let same = true;
+    forEach(newSub, (val, k) => {
+      if (!this.#curSub?.[k as Streamable].spec.equals(val.spec)) same = false;
+    });
+    return same;
+  }
+
+  #sendSpec(newSub?: Record<Streamable, Subscription>): void {
+    if (!newSub) return;
+    // Skip current sub and move to next only if sub has already been sent
+    if (this.#shouldSkip(newSub) && this.#syncSent) {
+      this.#advance();
+      return;
+    }
+
+    this.#numSyncs += 1;
+    const sync_id = this.#numSyncs.toString();
+    const spec = reduce(
+      newSub,
+      (spec, ent, k) => {
+        spec['known'][k] = ent.keyCache.known();
+        spec['subscribe'][k] = {
+          ...ent.spec.toWire(),
+          since: ent.keyCache.maxSeq(),
+        };
+        return spec;
+      },
+      { known: {}, subscribe: {}, sync_id: sync_id } as Record<string, StreamContent>,
+    );
+
+    this.#curSub = newSub;
+    this.#ws && this.#ws.send(JSON.stringify(spec));
```

Comment:
Let's use `this.#ws!.send()` instead.

The `!` communicates, "I know that this.#ws is defined by now even though typescript doesn't, and the code _should_ crash if I'm wrong because that's a bug".

By contrast, `this.#ws && this.#ws.send()` communicates, "I'm not sure if this.#ws is defined but if it is we should send the subscription".

(I'm assuming that if `#sendSpec()` doesn't actually send a spec, our Stream object is probably toast anyway)